pkg/services/sqlstore/migrations/accesscontrol/action_set_migration.go

```diff
@@ -0,0 +1,141 @@
+package accesscontrol
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"xorm.io/xorm"
+
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+)
+
+const AddActionSetMigrationID = "adding action set permissions"
+
+func AddActionSetPermissionsMigrator(mg *migrator.Migrator) {
+	mg.AddMigration(AddActionSetMigrationID, &actionSetMigrator{})
+}
+
+type actionSetMigrator struct {
+	sess     *xorm.Session
+	migrator *migrator.Migrator
+	migrator.MigrationBase
+}
+
+var _ migrator.CodeMigration = new(actionSetMigrator)
+
+func (m *actionSetMigrator) SQL(migrator.Dialect) string {
+	return "code migration"
+}
+
+func (m *actionSetMigrator) Exec(sess *xorm.Session, migrator *migrator.Migrator) error {
+	m.sess = sess
+	m.migrator = migrator
+	return m.addActionSetActions()
+}
+
+func (m *actionSetMigrator) addActionSetActions() error {
+	var results []accesscontrol.Permission
+
+	// Find action sets and dashboard permissions for managed roles
+	// We don't need all dashboard permissions, just enough to help us determine what action set permissions to add
+	sql := `
+	SELECT permission.role_id, permission.action, permission.scope FROM permission
+		LEFT JOIN role ON permission.role_id = role.id
+		WHERE permission.action IN ('dashboards:read', 'dashboards:write', 'dashboards.permissions:read', 'dashboards:view', 'dashboards:edit', 'dashboards:admin', 'folders:view', 'folders:edit', 'folders:admin')
+		AND role.name LIKE 'managed:%'
+`
+	if err := m.sess.SQL(sql).Find(&results); err != nil {
+		return fmt.Errorf("failed to query permissions: %w", err)
+	}
+
+	// group permissions by map[roleID]map[scope]actionSet
+	groupedPermissions := make(map[int64]map[string]string)
+	hasActionSet := make(map[int64]map[string]bool)
+	for _, result := range results {
+		// keep track of which dash/folder permission grants already have an action set permission
+		if isActionSetAction(result.Action) {
+			if _, ok := hasActionSet[result.RoleID]; !ok {
+				hasActionSet[result.RoleID] = make(map[string]bool)
+			}
+			hasActionSet[result.RoleID][result.Scope] = true
+			delete(groupedPermissions[result.RoleID], result.Scope)
+			continue
+		}
+
+		// don't add action set permissions where they already exist
+		if _, has := hasActionSet[result.RoleID]; has && hasActionSet[result.RoleID][result.Scope] {
+			continue
+		}
+
+		if _, ok := groupedPermissions[result.RoleID]; !ok {
+			groupedPermissions[result.RoleID] = make(map[string]string)
+		}
+
+		// store the most permissive action set permission
+		currentActionSet := groupedPermissions[result.RoleID][result.Scope]
+		switch result.Action {
+		case "dashboards:read":
```

Comment:
Does it work for folders as well? It seems like yes, but implicitly (if user has "dashboards:read" on folder "folder:uid:1" then it will get "view" on this, right?

Reply:
@alexanderzobnin i am looking at the testcases. @IevaVasiljeva they are well covered.

we could add one more with this example to make sure that we would get the "folders:view" for this one.